File: 1/Topics/linked_list/hackerrank_algorithms_linked_list_complete/linked_list.cxx
#include<iostream>
using namespace std;

class Node{
    public:
        int data;
        Node* list;
};

Node* head=NULL;

void print(Node* head){

    Node* temp=head;
    while(temp!=nullptr)
    {
        cout<<temp->data;
        temp=temp->list;
    }

}


void  insert(Node* x){
    Node* temp=head;
    if(head==NULL)
    head=x;
    else
    {
        while(temp->list!=NULL)
        temp=temp->list;
        temp->list=x;
    }
    print(head);
    cout<<endl;
}

    void insertbegin(Node* x){

        if(head==NULL)
        head=x;
        else
        {
            x->list=head->list;
            head=x;
        }
        print(head);
    }

int main(){
    Node* a=new Node();
    Node* b=new Node();
    Node* c=new Node();
    a->data=1;
    a->list=NULL;
    head=a;
    //cout<<head->data;
    b->data=2;
    b->list=NULL;
    insertbegin(b);
    c->data=3;
    c->list=NULL;
    insertbegin(c);
    return 0;
}

File: 2/strivers_sheet/day5/LinkedList.java
class Node{
    int data;
    Node next;

    Node(int data){
        this.data=data;
        this.next = null;
    }
}


public class LinkedList{

    Node head;

    public Node insert(int data) {
        if (head == null) {
            head = new Node(data);
            return head;
        }
        Node temp = head;
        while (temp.next != null) {
            temp = temp.next;
        }
        temp.next = new Node(data);
        return head;
    }

    public Node reverseLL(Node head){

        Node newHead = null;

        while(head!=null){
            Node next = head.next;
            head.next = newHead;
            newHead = head;
            head = next;
        }
        return newHead;

    }

    public Node middleLL(Node head){
        Node fast = head;
        Node slow = head;

        while(fast!=null && fast.next!=null){
            slow = slow.next;
            fast = fast.next.next;
        }
        return slow;
    }

    public void traverse(Node head) {
        Node temp = head;
        while (temp != null) {
            System.out.print(temp.data + " ");
            temp = temp.next;
        }
        System.out.println();
    }

    public Node insertFirst(Node n1){
        if(head==null){
            return n1;
        }
        else{
            n1.next = head;
            head = n1;
        }
        return head;
    }

    public void mergeTwoLL(LinkedList list1, Node head){
    
        

    }


    public static void main(String[] main){
        LinkedList list = new LinkedList();
        list.insert(1);
        list.insert(3);
        list.insert(5);

        LinkedList list2 = new LinkedList();
        list2.insert(0);
        list2.insert(2);
        list2.insert(4);

        
    }

}